From c43bfed8deaa0dab47c54db9b8f374853d345a6b Mon Sep 17 00:00:00 2001
From: Natanael Copa <ncopa@alpinelinux.org>
Date: Mon, 7 Dec 2015 10:41:13 +0000
Subject: [PATCH] db: add support for --no-cache

Implement --no-cache. The index is read directly from network and not
cached. This is useful for docker, where you install a set of packages
and directly after purge the cache. (see
https://github.com/gliderlabs/docker-alpine/blob/1fc9e59d1689fc4eaf930ec66389fe58062fccec/builder/scripts/apk-install)

fixes #4905
---
 src/apk.c         | 4 ++++
 src/apk_defines.h | 1 +
 src/database.c    | 8 +++++++-
 3 files changed, 12 insertions(+), 1 deletion(-)

diff --git a/src/apk.c b/src/apk.c
index 8632587..91673d8 100644
--- a/src/apk.c
+++ b/src/apk.c
@@ -124,6 +124,9 @@ static int option_parse_global(void *ctx, struct apk_db_options *dbopts, int opt
 	case 0x109:
 		apk_flags |= APK_NO_NETWORK;
 		break;
+	case 0x115:
+		apk_flags |= APK_NO_CACHE;
+		break;
 	case 0x112:
 		dbopts->arch = optarg;
 		break;
@@ -173,6 +176,7 @@ static const struct apk_option options_global[] = {
 	{ 0x108, "repositories-file", "Override repositories file",
 				required_argument, "REPOFILE" },
 	{ 0x109, "no-network",	"Do not use network (cache is still used)" },
+	{ 0x115, "no-cache",	"Read uncached index from network" },
 	{ 0x112, "arch",	"Use architecture with --root",
 				required_argument, "ARCH" },
 	{ 0x114, "print-arch",	"Print default arch and exit" },
diff --git a/src/apk_defines.h b/src/apk_defines.h
index 69113b4..2aca364 100644
--- a/src/apk_defines.h
+++ b/src/apk_defines.h
@@ -76,6 +76,7 @@ extern char **apk_argv;
 #define APK_NO_NETWORK		0x1000
 #define APK_OVERLAY_FROM_STDIN	0x2000
 #define APK_NO_SCRIPTS		0x4000
+#define APK_NO_CACHE		0x8000
 
 /* default architecture for APK packages. */
 #if defined(__x86_64__)
diff --git a/src/database.c b/src/database.c
index 8a56401..31ac3e4 100644
--- a/src/database.c
+++ b/src/database.c
@@ -2198,7 +2198,13 @@ int apk_db_add_repository(apk_database_t _db, apk_blob_t _repository)
 			if (apk_flags & APK_UPDATE_CACHE)
 				apk_repository_update(db, repo);
 		}
-		r = apk_repo_format_cache_index(APK_BLOB_BUF(buf), repo);
+		if (apk_flags & APK_NO_CACHE) {
+			r = apk_repo_format_real_url(db, repo, NULL, buf, sizeof(buf));
+			if (r == 0)
+				apk_message("fetch %s", buf);
+		} else {
+			r = apk_repo_format_cache_index(APK_BLOB_BUF(buf), repo);
+		}
 	} else {
 		db->local_repos |= BIT(repo_num);
 		db->available_repos |= BIT(repo_num);
-- 
2.6.3

